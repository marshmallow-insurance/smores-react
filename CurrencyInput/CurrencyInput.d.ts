import React, { FocusEvent, FormEvent } from 'react';
import { CommonFieldProps } from '../fields/commonFieldTypes';
export interface Props extends CommonFieldProps {
    placeholder: string;
    name?: string;
    value: string;
    decimalNumber?: boolean;
    onBlur?: (e: FocusEvent<HTMLInputElement>) => void;
    min?: number;
    max?: number;
}
/** on change or on input required */
type InputProps = {
    /** on change is required and on input optional */
    onChange: (e: string) => void;
    onInputChange?: (e: FormEvent<HTMLInputElement>) => void;
} | {
    /** on input is required and on change optional */
    onChange?: (e: string) => void;
    onInputChange: (e: FormEvent<HTMLInputElement>) => void;
};
export type CurrencyInputProps = Props & InputProps;
export declare const CurrencyInput: React.ForwardRefExoticComponent<CurrencyInputProps & React.RefAttributes<HTMLInputElement>>;
export {};
