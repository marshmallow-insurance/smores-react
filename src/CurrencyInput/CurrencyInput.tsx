import React, {
  FocusEvent,
  FormEvent,
  forwardRef,
  ForwardedRef,
} from 'react'

import { Box } from '../Box'
import { Field } from '../fields/Field'
import { CommonFieldProps } from '../fields/commonFieldTypes'

import { useUniqueId } from '../utils/id'
import {
  Input,
  StyledFrontIcon,
} from '../fields/components/CommonInput'

export interface Props extends CommonFieldProps {
  placeholder: string
  name?: string
  value: string
  onBlur?: (e: FocusEvent<HTMLInputElement>) => void
  min?: number
  max?: number
}

/** on change or on input required */
type InputProps =
  | {
      /** on change is required and on input optional */
      onChange: (e: string) => void
      onInputChange?: (e: FormEvent<HTMLInputElement>) => void
    }
  | {
      /** on input is required and on change optional */
      onChange?: (e: string) => void
      onInputChange: (e: FormEvent<HTMLInputElement>) => void
    }

export type CurrencyInputProps = Props & InputProps

export const CurrencyInput = forwardRef(function CurrencyInput(
  {
    id: idProp,
    placeholder,
    name,
    value,
    onChange,
    onInputChange,
    onBlur,
    min,
    max,
    disabled = false,
    error = false,
    trailingIcon,
    fallbackStyle,
    ...fieldProps
  }: CurrencyInputProps,
  ref: ForwardedRef<HTMLInputElement>,
) {
  const id = useUniqueId(idProp)
  const frontIcon = "pound-regular"

  const formatCurrency = (value: string) => {
    const decimalIndex = value.indexOf('.')
    if (decimalIndex >= 0 && value.length > decimalIndex + 1) {
      const fractionalString = value.substring(decimalIndex + 1).substring(0, 2)
      return `${value.substring(0, decimalIndex)}.${fractionalString}`
    } else {
      return value
    }
  }

  const applyMinMax = (value: string) => {
    const numberValue = Number(value)
    if(min && numberValue < min) return min.toString()
    if(max && numberValue > max) return max.toString()
    return value
  }

  const handleChange = (value: string) => {
    if (onChange === undefined) return
    const EMPTY_INPUT = ''

    if (value === EMPTY_INPUT) {
      onChange(value)
    } else {
      const amount = formatCurrency(value)
      const normalisedValue = applyMinMax(amount)

      onChange(normalisedValue)
    }
  }

  return (
    <Field {...fieldProps} htmlFor={id} error={error}>
      <Box flex alignItems="center" justifyContent="flex-start">
        <StyledFrontIcon
          disabled={disabled}
          render={frontIcon}
          color="sesame"
        />
        <Input
          ref={ref}
          error={error}
          disabled={disabled}
          type="number"
          id={id}
          name={name}
          placeholder={placeholder}
          value={value}
          frontIcon={frontIcon}
          fallbackStyle={fallbackStyle}
          onChange={(e: FormEvent<HTMLInputElement>) => {
            onChange && handleChange(e.currentTarget.value)
            onInputChange && onInputChange(e)
          }}
          onBlur={onBlur}
        />
      </Box>
    </Field>
  )
})
