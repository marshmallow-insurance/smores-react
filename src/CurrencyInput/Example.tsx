import React, { useState } from 'react'

import { CurrencyInput } from './CurrencyInput'

export const Example = () => {
  const [value, setValue] = useState('')

  return (
    <form>
      <CurrencyInput
        id="numberInput"
        label="Amount paid"
        name="numberInput"
        onChange={(e) => setValue(e)}
        placeholder="100.00"
        value={value}
        min={-200}
        max={200}
        required
      />
    </form>
  )
}
