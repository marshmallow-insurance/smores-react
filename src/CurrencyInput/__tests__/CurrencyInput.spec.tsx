import { render } from '@testing-library/react'
import React from 'react'

import { CurrencyInput } from '../CurrencyInput'

test('renders - number', () => {
  const { container } = render(
    <CurrencyInput
      id="CurrencyInput"
      label="Number"
      name="CurrencyInput"
      value=""
      onChange={(str) => console.log(str)}
      placeholder="0.00"
      required={true}
    />,
  )
  expect(container.firstChild).toMatchSnapshot()
})

test('renders - error', () => {
  const { container } = render(
    <CurrencyInput
      id="currencyInput"
      label="Currency"
      name="currencyInput"
      value="200"
      onChange={(str) => console.log(str)}
      placeholder="100.00"
      min={0}
      max={100}
      error={true}
      errorMsg="Oh gosh, this is an error isn't it!"
    />,
  )
  expect(container.firstChild).toMatchSnapshot()
})

test('renders - disabled', () => {
  const { container } = render(
    <CurrencyInput
      id="currencyInput"
      label="Currency"
      name="currencyInput"
      value=""
      onChange={(str) => console.log(str)}
      placeholder="100.00"
      disabled={true}
    />,
  )
  expect(container.firstChild).toMatchSnapshot()
})
