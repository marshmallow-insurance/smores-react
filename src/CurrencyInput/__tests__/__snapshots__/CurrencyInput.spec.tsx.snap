// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders - disabled 1`] = `
<div
  class="sc-aXZVg sc-gsFSXq byxNQM cZVmFI"
>
  <label
    class="sc-aXZVg sc-fqkvVR beCtdt loBhWK"
    color="sesame"
    cursor="inherit"
    for="currencyInput"
    title=""
  >
    Currency
  </label>
  <div
    class="sc-aXZVg byxNQM"
  >
    <div
      class="sc-aXZVg cowLnB"
    >
      <span
        class="sc-aXZVg sc-gEvEer byxNQM eCTqKS sc-imWYAI cPbbMy"
        color="sesame"
        disabled=""
        rotate="0"
        size="24"
      >
        <svg
          fill="none"
          height="100%"
          viewBox="0 0 24 24"
          width="100%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.99 15.876l1.669.168c0 .685-.154 1.322-.462 1.91-.3.587-.77 1.077-1.406 1.469-.637.385-1.382.577-2.235.577H6v-1.637c.784-.392 1.43-.983 1.941-1.773.518-.791.777-1.617.777-2.477 0-.469-.07-.983-.21-1.543H6.063v-1.616h1.91a45.567 45.567 0 01-.388-1.102 8.872 8.872 0 01-.263-1.038 6.013 6.013 0 01-.126-1.176c0-.874.224-1.668.672-2.382a4.6 4.6 0 011.794-1.658A5.12 5.12 0 0112.107 3c1.456 0 2.6.378 3.432 1.133.832.756 1.273 1.735 1.322 2.939l-1.69.293a3.215 3.215 0 00-.178-.955 2.874 2.874 0 00-.472-.87c-.21-.28-.518-.5-.924-.662-.405-.168-.884-.251-1.437-.251-.889 0-1.637.293-2.246.881-.601.588-.902 1.319-.902 2.193 0 .126.007.26.02.399l.043.367c.02.105.052.245.094.42.05.175.084.304.105.388.021.077.067.22.136.43.07.21.112.343.126.4.021.048.074.195.158.44.084.238.133.374.147.41h4.712v1.615h-4.209c.133.511.2.994.2 1.448 0 1.784-.71 3.233-2.13 4.345h4.91c.518 0 .963-.077 1.333-.23.371-.162.648-.375.83-.641.181-.266.31-.528.388-.787.077-.266.115-.542.115-.83z"
            fill="#636768"
          />
        </svg>
      </span>
      <input
        class="sc-kAyceB gBcClo"
        disabled=""
        id="currencyInput"
        name="currencyInput"
        placeholder="100.00"
        type="number"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`renders - error 1`] = `
<div
  class="sc-aXZVg sc-gsFSXq byxNQM cZVmFI"
>
  <label
    class="sc-aXZVg sc-fqkvVR beCtdt ikzCDy"
    color="strawberry"
    cursor="inherit"
    for="currencyInput"
    title=""
  >
    Currency
  </label>
  <div
    class="sc-aXZVg byxNQM"
  >
    <div
      class="sc-aXZVg cowLnB"
    >
      <span
        class="sc-aXZVg sc-gEvEer byxNQM eCTqKS sc-imWYAI WoYwI"
        color="sesame"
        rotate="0"
        size="24"
      >
        <svg
          fill="none"
          height="100%"
          viewBox="0 0 24 24"
          width="100%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.99 15.876l1.669.168c0 .685-.154 1.322-.462 1.91-.3.587-.77 1.077-1.406 1.469-.637.385-1.382.577-2.235.577H6v-1.637c.784-.392 1.43-.983 1.941-1.773.518-.791.777-1.617.777-2.477 0-.469-.07-.983-.21-1.543H6.063v-1.616h1.91a45.567 45.567 0 01-.388-1.102 8.872 8.872 0 01-.263-1.038 6.013 6.013 0 01-.126-1.176c0-.874.224-1.668.672-2.382a4.6 4.6 0 011.794-1.658A5.12 5.12 0 0112.107 3c1.456 0 2.6.378 3.432 1.133.832.756 1.273 1.735 1.322 2.939l-1.69.293a3.215 3.215 0 00-.178-.955 2.874 2.874 0 00-.472-.87c-.21-.28-.518-.5-.924-.662-.405-.168-.884-.251-1.437-.251-.889 0-1.637.293-2.246.881-.601.588-.902 1.319-.902 2.193 0 .126.007.26.02.399l.043.367c.02.105.052.245.094.42.05.175.084.304.105.388.021.077.067.22.136.43.07.21.112.343.126.4.021.048.074.195.158.44.084.238.133.374.147.41h4.712v1.615h-4.209c.133.511.2.994.2 1.448 0 1.784-.71 3.233-2.13 4.345h4.91c.518 0 .963-.077 1.333-.23.371-.162.648-.375.83-.641.181-.266.31-.528.388-.787.077-.266.115-.542.115-.83z"
            fill="#636768"
          />
        </svg>
      </span>
      <input
        class="sc-kAyceB eGEDsu"
        id="currencyInput"
        name="currencyInput"
        placeholder="100.00"
        type="number"
        value="200"
      />
    </div>
  </div>
  <span
    class="sc-aXZVg sc-fqkvVR cMpUSh eXBKnw"
    color="strawberry"
    cursor="inherit"
    title=""
  >
    Oh gosh, this is an error isn't it!
  </span>
</div>
`;

exports[`renders - number 1`] = `
<div
  class="sc-aXZVg sc-gsFSXq byxNQM cZVmFI"
>
  <label
    class="sc-aXZVg sc-fqkvVR beCtdt loBhWK"
    color="sesame"
    cursor="inherit"
    for="CurrencyInput"
    title=""
  >
    Number
    <span
      class="sc-aXZVg sc-fqkvVR byxNQM gsSbBT"
      color="strawberry"
      cursor="inherit"
      title=""
    >
      *
    </span>
  </label>
  <div
    class="sc-aXZVg byxNQM"
  >
    <div
      class="sc-aXZVg cowLnB"
    >
      <span
        class="sc-aXZVg sc-gEvEer byxNQM eCTqKS sc-imWYAI WoYwI"
        color="sesame"
        rotate="0"
        size="24"
      >
        <svg
          fill="none"
          height="100%"
          viewBox="0 0 24 24"
          width="100%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.99 15.876l1.669.168c0 .685-.154 1.322-.462 1.91-.3.587-.77 1.077-1.406 1.469-.637.385-1.382.577-2.235.577H6v-1.637c.784-.392 1.43-.983 1.941-1.773.518-.791.777-1.617.777-2.477 0-.469-.07-.983-.21-1.543H6.063v-1.616h1.91a45.567 45.567 0 01-.388-1.102 8.872 8.872 0 01-.263-1.038 6.013 6.013 0 01-.126-1.176c0-.874.224-1.668.672-2.382a4.6 4.6 0 011.794-1.658A5.12 5.12 0 0112.107 3c1.456 0 2.6.378 3.432 1.133.832.756 1.273 1.735 1.322 2.939l-1.69.293a3.215 3.215 0 00-.178-.955 2.874 2.874 0 00-.472-.87c-.21-.28-.518-.5-.924-.662-.405-.168-.884-.251-1.437-.251-.889 0-1.637.293-2.246.881-.601.588-.902 1.319-.902 2.193 0 .126.007.26.02.399l.043.367c.02.105.052.245.094.42.05.175.084.304.105.388.021.077.067.22.136.43.07.21.112.343.126.4.021.048.074.195.158.44.084.238.133.374.147.41h4.712v1.615h-4.209c.133.511.2.994.2 1.448 0 1.784-.71 3.233-2.13 4.345h4.91c.518 0 .963-.077 1.333-.23.371-.162.648-.375.83-.641.181-.266.31-.528.388-.787.077-.266.115-.542.115-.83z"
            fill="#636768"
          />
        </svg>
      </span>
      <input
        class="sc-kAyceB btLkxU"
        id="CurrencyInput"
        name="CurrencyInput"
        placeholder="0.00"
        type="number"
        value=""
      />
    </div>
  </div>
</div>
`;
